File: src/learn-topic-wise/01-strings/cs/001-min-move-to-make-palindrome.cs
// export const url = "[](https://leetcode.com/problems/minimum-number-of-moves-to-make-palindrome/description/)"

namespace learning_dsa_csharp._01_strings_arrays_hash._001_min_move_to_make_palindrome
{
    internal class Solution
    {
        public int MinMovesToMakePalindrome(string s)
        {
            char[] cStr = s.ToCharArray();

            int l = 0;
            int r = cStr.Length - 1;

            int count = 0;
            int center_i = -1;

            while (l < r)
            {
                if (cStr[l] == cStr[r])
                {
                    l++;
                    r--;
                    continue;
                }

                int k = l + 1;
                for (; k < r; k++)
                {
                    if (cStr[k] == cStr[r])
                    {
                        break;
                    }
                }

                if (k == r)
                {
                    center_i = r;
                    r--;
                    continue;
                }

                for (int j = k; j > l; j--)
                {
                    cStr = SwapChar(cStr, j, j - 1);
                    count++;
                }

                l++;
                r--;
            }

            if (center_i != -1)
            {
                count += (center_i - cStr.Length / 2);
            }

            return count;
        }

        public char[] SwapChar(char[] s, int i1, int i2)
        {
            char t = s[i1];
            s[i1] = s[i2];
            s[i2] = t;

            return s;
        }
    }
}

File: src/learn-topic-wise/01-strings/hard/001-min-move-to-make-palindrome.ts
export const url =
    '[min move to make palindrome](https://leetcode.com/problems/minimum-number-of-moves-to-make-palindrome/description/)';

function minMovesToMakePalindrome(s: string): number {
    const sArr = [...s];

    let l = 0,
        r = s.length - 1,
        res = 0;
    while (l < r) {
        if (sArr[l] === sArr[r]) {
            ++l;
            --r;
        } else {
            let t = r;
            while (t > l && sArr[t] !== sArr[l]) {
                --t;
            }

            // sArr[l] is likely middle char
            if (t === l) {
                // just swap once, +1 move, but l & r remains same
                [sArr[t], sArr[t + 1]] = [sArr[t + 1], sArr[t]];
                ++res;
            } else {
                // swap and increment moves counter
                while (t < r) {
                    // swaps all the way till `r` and sArr[r] as well
                    [sArr[t], sArr[t + 1]] = [sArr[t + 1], sArr[t]];

                    ++t;
                    ++res;
                }

                ++l;
                --r;
            }
        }
    }

    return res;
}

var input = 'aabb';
console.log(minMovesToMakePalindrome(input));
